### PR TITLE
fix(xbind): [iOS/macOS] Fix memory leak when using x:Bind

### PR DESCRIPTION
**Pull request by jeromelaban**
GitHub Issue (If applicable): fixes https://github.com/unoplatform/uno/issues/3914

## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the new behavior?

Using x:Bind on iOS/macOS will not keep UIElement instances alive.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: build/uno.winui.targets [hunk 1]
```diff
@@ -26,6 +26,7 @@
 			<UnoDefineConstants Condition="'$(MSBuildThisFile)'=='uno.winui.targets'">$(UnoDefineConstants);HAS_UNO_WINUI</UnoDefineConstants>
 			<UnoDefineConstants>$(UnoDefineConstants);UNO_HAS_FRAMEWORKELEMENT_MEASUREOVERRIDE</UnoDefineConstants>
 			<UnoDefineConstants>$(UnoDefineConstants);UNO_HAS_NO_IDEPENDENCYOBJECT</UnoDefineConstants>
+			<UnoDefineConstants Condition="'$(TargetFramework)'=='xamarinios10' or '$(TargetFramework)'=='xamarinmac20'">$(UnoDefineConstants);UNO_HAS_UIELEMENT_IMPLICIT_PINNING</UnoDefineConstants>
```
Comment (MatFillion) [inline build/uno.winui.targets:29]: UNO_HAS_UIELEMENT_IMPLICIT_PINNING  -> Do we have a table with all those constants and what they mean, when they're used, etc?

Comment (jeromelaban) [reply build/uno.winui.targets:29]: Done, see Uno.CrossTargetting.props.

File: src/Uno.CrossTargetting.props [hunk 2]
```diff
@@ -41,6 +41,19 @@
...
+  - __SKIA__: Used when building for the Skia backend
+  - __WASM__: Used when building for the WebAssembly backend
+  - UNO_REFERENCE_API: Used when the netstandard 2.0 Reference API is used.
+  - HAS_EXPENSIVE_TRYFINALLY: Used on WebAssembly until exceptions are supported by the WebAssembly specification
+  - __NETSTD_REFERENCE__: Used when building the reference assemblies for the netstandard 2.0 target
+  - UNO_HAS_UIELEMENT_IMPLICIT_PINNING: Used to mark targets that have specific constraints on
+										UIElements. On iOS, this means additional weak references
+										backed fields to handle opaque native reference pinning and avoid
+										memory leaks.
+  -->
```
Comment (davidjohnoliver) [inline src/Uno.CrossTargetting.props:55]:
```suggestion
  - Constants for Xamarin backends and SDK versions: https://docs.microsoft.com/en-us/xamarin/cross-platform/app-fundamentals/building-cross-platform-applications/platform-divergence-abstraction-divergent-implementation#conditional-compilation
  -->
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Controls/XamlEvent_Leak_UserControl_xBind.xaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:UITests.Shared.Windows_UI_Xaml.FrameworkElementTests"
+    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+    mc:Ignorable="d"
+    d:DesignHeight="300"
+    d:DesignWidth="400">
+
+    <Grid>
+		<!-- A named button with an event must not leak -->
```
Comment (davidjohnoliver) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Controls/XamlEvent_Leak_UserControl_xBind.xaml:13]:
```suggestion
		<!-- A named button with an x:Bind must not leak -->
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Controls/XamlEvent_Leak_UserControl_xBind_Event.xaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:UITests.Shared.Windows_UI_Xaml.FrameworkElementTests"
+    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+    mc:Ignorable="d"
+    d:DesignHeight="300"
+    d:DesignWidth="400">
+
+    <Grid>
+		<!-- A named button with an event must not leak -->
```
Comment (davidjohnoliver) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Controls/XamlEvent_Leak_UserControl_xBind_Event.xaml:13]:
```suggestion
		<!-- A named button with an x:Bind event must not leak -->
```

**Review by davidjohnoliver — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/Data/Binding.cs [hunk 5]
```diff
@@ -22,6 +22,19 @@ public partial class Binding : BindingBase
...
+		/// <summary>
+		/// On platforms which perform implicit and opaque pinning of native references, it
+		/// is required to keep weak references to direct and indirect references to UIElement instances.
+		/// Keeping weak references is costly,  so it's enabled only on select platforms.
+		/// Note that this is needed only for x:Bind related operations, where the lifespan of
+		/// the binding is explicitly tied to the object containing the weak references. This means
+		/// that there's weak references will be kept alive properly.
+		/// </summary>
+
+		private WeakReference _xBindSelector, _xBindBack, _compiledSource;
```
[the remark this answers is not in the document]
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Data/Binding.cs:35]: True. for `_xBindSelector`, `_xBindBack` it's not useful as they cannot be returned back to the pool, but `_compiledSource` would be useful indeed.

**Comment by jeromelaban** [comment]
@mergifyio backport release/stable/3.0

**Pull request merged**